plugin/pkg/shoot/quotavalidator/admission.go

```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
+		}
 	}
 
 	return nil
 }
+
+func (h *RejectShootIfQuotaExceeded) getShootsQuotaReferences(shoot garden.Shoot) ([]v1.ObjectReference, error) {
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "CrossSecretBinding":
+		usedCrossSB, err := h.crossSBLister.
+			CrossSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		return usedCrossSB.Quotas, nil
+	case "PrivateSecretBinding":
+		usedPrivateSB, err := h.privateSBLister.
+			PrivateSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		return usedPrivateSB.Quotas, nil
+	}
+	return nil, fmt.Errorf("Unknown binding type %s", shoot.Spec.Cloud.SecretBindingRef.Kind)
+}
+
+func (h *RejectShootIfQuotaExceeded) isQuotaExceeded(shoot garden.Shoot, quota garden.Quota) (*[]v1.ResourceName, error) {
+	allocatedResources, err := h.determineAllocatedResources(quota, shoot)
+	if err != nil {
+		return nil, err
+	}
+	requiredResources, err := h.determineRequiredResources(allocatedResources, shoot)
+	if err != nil {
+		return nil, err
+	}
+
+	exceededMetrics := make([]v1.ResourceName, 0)
+	for _, metric := range quotaMetricNames {
+		if _, ok := quota.Spec.Metrics[metric]; !ok {
+			continue
+		}
+		if !hasSufficientQuota(quota.Spec.Metrics[metric], requiredResources[metric]) {
+			exceededMetrics = append(exceededMetrics, metric)
+		}
+	}
+	if len(exceededMetrics) != 0 {
+		return &exceededMetrics, nil
+	}
+	return nil, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) determineAllocatedResources(quota garden.Quota, shoot garden.Shoot) (v1.ResourceList, error) {
+	shoots, err := h.findShootsReferQuota(quota, shoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect the resources which are allocated according to the shoot specs
+	allocatedResources := make(v1.ResourceList)
+	for _, s := range shoots {
+		shootResources, err := h.getShootResources(s)
+		if err != nil {
+			return nil, err
+		}
+		for _, metric := range quotaMetricNames {
+			allocatedResources[metric] = sumQuantity(allocatedResources[metric], shootResources[metric])
+		}
+	}
+
+	// TODO: We have to determine and add the amount of storage, which is allocated by manually created persistent volumes
+	// and the count of loadbalancer, which are created due to manually created services of type loadbalancer
+
+	return allocatedResources, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) findShootsReferQuota(quota garden.Quota, shoot garden.Shoot) ([]garden.Shoot, error) {
+	var shootsReferQuota []garden.Shoot
+
+	privateSecretBindings, err := h.getPrivateSecretBindingsReferQuota(quota)
+	if err != nil {
+		return nil, err
+	}
+	crossSecretBindings, err := h.getCrossSecretBindingsReferQuota(quota)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all shoots which are referencing found PrivateSecretBindings
+	for _, binding := range privateSecretBindings {
+		shoots, err := h.shootLister.
```

Comment:
Can we write these statements into one line (line breakings are not used anywhere else in the code)?

Reply:
done